File: frontend/src/api/offline.ts
/**
 * Offline Area Manager — `/imagery/offline/*` + `/imagery/usage`.
 *
 * ★ The pre-cache runs SERVER-SIDE now: the backend enumerates the AOI's XYZ tiles,
 *   checks its own cache, enforces the request budgets (LE_MAPBOX_MAX_*), downloads with
 *   bounded concurrency through the same read-through cache the live map uses, and writes
 *   the offline manifest. The client's job is the polygon, the zoom band, and honest
 *   display of the server's estimate/progress — not the download loop it used to run
 *   (which could not see the cache, the budgets, or survive a page navigation).
 */

import { fetchJson } from './client';

/** One AOI vertex, `(lon, lat)` — GeoJSON axis order. */
export type LonLat = [number, number];

export type BasemapKindStr = 'satellite' | 'hybrid' | 'terrain';

export interface OfflineAreaRequest {
  provider: string;
  kind: BasemapKindStr;
  polygon: LonLat[];
  zoom_min: number;
  zoom_max: number;
  project_id?: string | null;
}

export interface PrecacheBudget {
  per_operation_limit: number | null;
  max_prefetch_tiles: number | null;
  per_day_limit: number | null;
  per_day_used: number;
  within_budget: boolean;
  reason: string | null;
}

export interface OfflineEstimate {
  provider: string;
  zoom_min: number;
  zoom_max: number;
  total_tiles: number;
  cached_tiles: number;
  missing_tiles: number;
  estimated_upstream_requests: number;
  average_tile_size_bytes: number;
  /** `measured` = sampled from this machine's own cache; `default` = stated constant. */
  average_tile_size_source: 'measured' | 'default';
  estimated_storage_bytes: number;
  capped: boolean;
  budget: PrecacheBudget;
}

export type PrecacheState = 'pending' | 'running' | 'paused' | 'completed' | 'cancelled' | 'failed';

export interface PrecacheOperation {
  id: string;
  state: PrecacheState;
  provider: string;
  kind: BasemapKindStr;
  zoom_min: number;
  zoom_max: number;
  project_id: string | null;
  total_tiles: number;
  completed_tiles: number;
  skipped_cached: number;
  no_imagery_tiles: number;
  failed_tiles: number;
  remaining_tiles: number;
  downloaded_bytes: number;
  percent: number;
  started_at: string | null;
  finished_at: string | null;
  note: string | null;
  manifest_id: string | null;
}

export interface OfflineManifest {
  id: string;
  project_id: string | null;
  provider: string;
  kind: BasemapKindStr;
  cache_variant: string;
  polygon: LonLat[];
  bbox: [number, number, number, number];
  zoom_min: number;
  zoom_max: number;
  tile_size_px: number;
  tile_count: number;
  completed_tiles: number;
  no_imagery_tiles: number;
  missing_tiles: number;
  downloaded_bytes: number;
  created_at: string;
  expires_at: string | null;
}

export interface OfflineCoverage {
  provider: string;
  total_tiles: number;
  cached_tiles: number;
  coverage_ratio: number;
  by_zoom: Record<string, { total: number; cached: number }>;
  capped: boolean;
}

export const offlineApi = {
  estimate: (body: OfflineAreaRequest, signal?: AbortSignal): Promise<OfflineEstimate> =>
    fetchJson<OfflineEstimate>('/imagery/offline/estimate', { method: 'POST', body, signal }),

  coverage: (body: OfflineAreaRequest, signal?: AbortSignal): Promise<OfflineCoverage> =>
    fetchJson<OfflineCoverage>('/imagery/offline/coverage', { method: 'POST', body, signal }),

  start: (body: OfflineAreaRequest): Promise<PrecacheOperation> =>
    fetchJson<PrecacheOperation>('/imagery/offline/operations', { method: 'POST', body }),

  listOperations: (signal?: AbortSignal): Promise<PrecacheOperation[]> =>
    fetchJson<PrecacheOperation[]>('/imagery/offline/operations', { signal }),

  getOperation: (id: string, signal?: AbortSignal): Promise<PrecacheOperation> =>
    fetchJson<PrecacheOperation>(`/imagery/offline/operations/${id}`, { signal }),

  pause: (id: string): Promise<PrecacheOperation> =>
    fetchJson<PrecacheOperation>(`/imagery/offline/operations/${id}/pause`, { method: 'POST' }),

  resume: (id: string): Promise<PrecacheOperation> =>
    fetchJson<PrecacheOperation>(`/imagery/offline/operations/${id}/resume`, { method: 'POST' }),

  cancel: (id: string): Promise<PrecacheOperation> =>
    fetchJson<PrecacheOperation>(`/imagery/offline/operations/${id}/cancel`, { method: 'POST' }),

  retry: (id: string): Promise<PrecacheOperation> =>
    fetchJson<PrecacheOperation>(`/imagery/offline/operations/${id}/retry`, { method: 'POST' }),

  listManifests: (signal?: AbortSignal): Promise<OfflineManifest[]> =>
    fetchJson<OfflineManifest[]>('/imagery/offline/manifests', { signal }),

  deleteManifest: (id: string): Promise<void> =>
    fetchJson<void>(`/imagery/offline/manifests/${id}`, { method: 'DELETE', parse: 'none' }),
};

// ── automatic viewport caching ────────────────────────────────────────────────

/** A settled viewport, in PROVIDER zoom levels (map zoom minus the 512px zoomShift). */
export interface ViewportReport {
  provider: string;
  kind: BasemapKindStr;
  west: number;
  south: number;
  east: number;
  north: number;
  zoom: number;
  project_id?: string | null;
}

export interface AutoCacheSession {
  started_at: string;
  project_id: string | null;
  tiles_requested: number;
  tiles_downloaded: number;
  tiles_skipped_cached: number;
  tiles_no_imagery: number;
  tiles_failed: number;
  bytes_downloaded: number;
  limit_reached: string | null;
}

export interface AutoCacheStatus {
  enabled: boolean;
  online: boolean;
  active: boolean;
  provider: string | null;
  kind: BasemapKindStr;
  zoom: number | null;
  viewport_tiles: number;
  cached_tiles: number;
  missing_tiles: number;
  coverage_percent: number | null;
  queued_tiles: number;
  /** More missing tiles remain beyond this chunk — re-report on drain to continue. */
  truncated: boolean;
  note: string | null;
  session: AutoCacheSession;
}

export const autoCacheApi = {
  /** Report a settled viewport — queues background caching, returns coverage. */
  reportViewport: (body: ViewportReport, signal?: AbortSignal): Promise<AutoCacheStatus> =>
    fetchJson<AutoCacheStatus>('/imagery/cache/viewport', { method: 'POST', body, signal }),

  getStatus: (signal?: AbortSignal): Promise<AutoCacheStatus> =>
    fetchJson<AutoCacheStatus>('/imagery/cache/status', { signal }),

  setEnabled: (enabled: boolean): Promise<AutoCacheStatus> =>
    fetchJson<AutoCacheStatus>('/imagery/cache/enabled', { method: 'PUT', body: { enabled } }),
};

/** Human-readable byte size for estimates ("1.7 GB"). */
export function formatBytes(bytes: number): string {
  if (!Number.isFinite(bytes) || bytes < 0) return '—';
  if (bytes < 1024) return `${bytes} B`;
  const units = ['KB', 'MB', 'GB', 'TB'];
  let value = bytes;
  let unit = -1;
  do {
    value /= 1024;
    unit += 1;
  } while (value >= 1024 && unit < units.length - 1);
  return `${value.toFixed(value >= 100 ? 0 : 1)} ${units[unit]}`;
}

File: backend/app/schemas/offline.py
"""Offline area pre-caching — the wire types for ``/imagery/offline/*`` (Offline Area Manager).

★ **The pre-cache and the map share ONE cache.** These endpoints drive the same
read-through tile cache the proxy writes; "download an area" is enumerating its XYZ tiles
and pulling each one through the normal path. The estimate/coverage answers therefore
describe the exact bytes the offline map will later read.

★ Estimates are labelled with their PROVENANCE (``average_tile_size_source``): a measured
mean from this operator's own cache beats any constant, but when nothing is cached yet the
default is stated as a default — never presented as a measurement.
"""

from __future__ import annotations

from datetime import datetime
from typing import Annotated, Literal

from pydantic import Field, model_validator

from .common import ApiModel
from .imagery import BasemapKind

__all__ = [
    "OfflineAreaRequest",
    "OfflineCoverage",
    "OfflineEstimate",
    "OfflineManifest",
    "PrecacheBudget",
    "PrecacheOperation",
    "PrecacheStartRequest",
]

#: One polygon vertex, ``(lon, lat)`` — GeoJSON axis order, like ``BBox``.
LonLat = Annotated[list[float], Field(min_length=2, max_length=2)]


class OfflineAreaRequest(ApiModel):
    """An AOI + zoom band + provider — the input to estimate/coverage/start."""

    provider: str = Field(min_length=1, description="Registry key, e.g. mapbox_satellite.")
    kind: BasemapKind = "satellite"
    polygon: Annotated[list[LonLat], Field(min_length=3)] = Field(
        description="AOI ring as (lon, lat) vertices; a bounding box is its 4 corners."
    )
    zoom_min: int = Field(ge=0, le=24)
    zoom_max: int = Field(ge=0, le=24)
    project_id: str | None = Field(
        default=None, description="Optional project this offline area belongs to."
    )

    @model_validator(mode="after")
    def _zoom_ordering(self) -> "OfflineAreaRequest":
        if self.zoom_min > self.zoom_max:
            raise ValueError("zoom_min must not exceed zoom_max.")
        return self


class PrecacheStartRequest(OfflineAreaRequest):
    """Start a download. Identical to the estimate input — what you saw is what runs."""


class PrecacheBudget(ApiModel):
    """The configured request budgets, evaluated against this request."""

    per_operation_limit: int | None = Field(
        default=None, description="LE_*_MAX_TILE_REQUESTS_PER_OPERATION; null = unlimited."
    )
    max_prefetch_tiles: int | None = Field(
        default=None, description="LE_*_MAX_PREFETCH_TILES; null = unlimited."
    )
    per_day_limit: int | None = Field(
        default=None, description="LE_*_MAX_TILE_REQUESTS_PER_DAY; null = unlimited."
    )
    per_day_used: int = Field(
        default=0, description="Upstream requests already recorded today (persisted)."
    )
    within_budget: bool
    reason: str | None = Field(
        default=None, description="Why the request exceeds a budget, naming the setting."
    )


class OfflineEstimate(ApiModel):
    """What a pre-cache run would cost — tiles, upstream requests, storage."""

    provider: str
    zoom_min: int
    zoom_max: int
    total_tiles: int
    cached_tiles: int = Field(description="Already present (fresh) in the tile cache.")
    missing_tiles: int = Field(description="What a download would actually fetch.")
    estimated_upstream_requests: int = Field(
        description="= missing_tiles: one request per tile (@2x tiles are ONE request)."
    )
    average_tile_size_bytes: int
    average_tile_size_source: Literal["measured", "default"] = Field(
        description="measured = sampled from this cache; default = stated constant."
    )
    estimated_storage_bytes: int
    capped: bool = Field(
        description="True when enumeration stopped at the hard cap — shrink the AOI."
    )
    budget: PrecacheBudget


class PrecacheOperation(ApiModel):
    """One pre-cache run's live status."""

    id: str
    state: Literal["pending", "running", "paused", "completed", "cancelled", "failed"]
    provider: str
    kind: BasemapKind
    zoom_min: int
    zoom_max: int
    project_id: str | None = None
    total_tiles: int
    completed_tiles: int
    skipped_cached: int = Field(description="Already in cache — never went upstream.")
    no_imagery_tiles: int = Field(description="Provider has nothing there (ocean/gap).")
    failed_tiles: int
    remaining_tiles: int
    downloaded_bytes: int
    percent: float = Field(ge=0.0, le=100.0)
    started_at: datetime | None = None
    finished_at: datetime | None = None
    note: str | None = Field(
        default=None, description="Why the run paused/failed, when it did."
    )
    manifest_id: str | None = Field(
        default=None, description="Set once the manifest is written (completed runs)."
    )


class OfflineManifest(ApiModel):
    """A durable record of a downloaded offline area.

    ★ Enough to answer, later and offline: which provider and configuration produced
    these tiles, over what AOI and zoom band, how complete the area is, and when the
    cached imagery expires. ``cache_variant`` is the provider's config fingerprint — the
    same string in the tile cache key, so a config change is visible as a mismatch.
    """

    id: str
    project_id: str | None = None
    provider: str
    kind: BasemapKind
    cache_variant: str
    polygon: list[LonLat]
    bbox: Annotated[list[float], Field(min_length=4, max_length=4)] = Field(
        description="(min_lon, min_lat, max_lon, max_lat) of the AOI."
    )
    zoom_min: int
    zoom_max: int
    tile_size_px: int
    tile_count: int
    completed_tiles: int
    no_imagery_tiles: int
    missing_tiles: int
    downloaded_bytes: int
    created_at: datetime
    expires_at: datetime | None = Field(
        default=None,
        description="created_at + the provider's effective cache TTL; null = no expiry.",
    )


class OfflineCoverage(ApiModel):
    """How much of an AOI the cache currently holds, per zoom and overall."""

    provider: str
    total_tiles: int
    cached_tiles: int
    coverage_ratio: float = Field(ge=0.0, le=1.0)
    by_zoom: dict[str, dict[str, int]] = Field(
        default_factory=dict,
        description='Per zoom: {"15": {"total": n, "cached": m}, ...}.',
    )
    capped: bool = False
